api/pin.go

```diff
@@ -0,0 +1,426 @@
+		return
+	}
+
+	isRawInDB, err := p.db.IsPinnedFileRaw(chunk.Address(addr))
+	if err != nil {
+		log.Error("Root hash is not pinned" + err.Error())
+		return
+	}
+
+	// Walk the root hash and unpin all the chunks
+	walkerFunction := func(ref storage.Reference) error {
+		chunkAddr := p.removeDecryptionKeyFromChunkHash(ref)
+		err := p.db.Set(context.TODO(), chunk.ModeSetUnpin, chunkAddr)
+		if err != nil {
+			log.Error("Could not unpin chunk. Address " + hex.EncodeToString(chunkAddr))
+			return err
+		} else {
+			log.Trace("Unpinning chunk", "Address", fmt.Sprintf("%0x", chunkAddr))
+		}
+		return nil
+	}
+	p.walkChunksFromRootHash(rootHash, isRawInDB, credentials, walkerFunction)
+
+	// Check if the root chunk exists in pinIndex
+	// If it is not.. then the pin counter became 0
+	// so remove the root hash from pinFilesIndex
+	isRootChunkPinned := p.db.IsChunkPinned(chunk.Address(p.removeDecryptionKeyFromChunkHash(addr)))
+	if !isRootChunkPinned {
+		err := p.db.Set(context.TODO(), chunk.ModeSetUnpinFile, addr)
+		if err != nil {
+			// TODO: if this happens, we should go back and revert the entire file's chunks
+			log.Error("Could not unpin root chunk. Address " + fmt.Sprintf("%0x", addr))
+		}
+	}
+}
+
+// ListPinFiles functions logs information of all the files that are pinned
+// in the current local node. It displays the root hash of the pinned file
+// or collection. It also display two vital information
+//     1) Size of the pinned file or collection
+//     2) the number of times that particular file or collection is pinned.
+func (p *PinAPI) ListPinFiles() {
+	pinnedFiles := p.db.GetPinFilesIndex()
+	for k, v := range pinnedFiles {
+
+		addr, err := hex.DecodeString(k)
+		if err != nil {
+			log.Error("Error decoding root hash" + err.Error())
+			return
+		}
+		pinCounter, err := p.db.GetPinCounterOfChunk(chunk.Address(p.removeDecryptionKeyFromChunkHash(addr)))
+
+		isRaw := false
+		if v > 0 {
+			isRaw = true
+		}
+		noOfChunks := p.getNoOfChunks(k, isRaw, "")
+
+		log.Info("Pinned file", "Address", k, "NoOfChunks", noOfChunks, "pinCounter", pinCounter)
+	}
+}
+
+// LogPinnedChunks logs all the chunks that are pinned as part of the given root hash.
+// This is mostly used for debugging.
+func (p *PinAPI) LogPinnedChunks(rootHash string, credentials string) {
+
+	addr, err := hex.DecodeString(rootHash)
+	if err != nil {
+		log.Error("Error decoding root hash" + err.Error())
+		return
+	}
+
+	isRawInDB, err := p.db.IsPinnedFileRaw(chunk.Address(addr))
+	if err != nil {
+		log.Error("Root hash is not pinned" + err.Error())
+		return
+	}
+
+	walkerFunction := func(ref storage.Reference) error {
+		log.Info("Chunk", "Address", fmt.Sprintf("%0x", ref))
+		return nil
+	}
+	p.walkChunksFromRootHash(rootHash, isRawInDB, credentials, walkerFunction)
+}
+
+func (p *PinAPI) walkChunksFromRootHash(rootHash string, isRaw bool, credentials string, executeFunc func(storage.Reference) error) {
```

Comment:
can you please document what this function is supposed to do?
more importantly - this function is not unit tested.
i'd like to see a test that adds data to the localstore, including a manifest that wraps the content in some way, and that we can surely assert that this function spits out chunk references and traverses everything - from the root chunk to the intermediate manifest chunks. i.e. it should show everything in the localstore after the time of the data insertion. i'd also ask for more elaborate manifests in this case (just generate a big manifest that spans over different chunks and that it has sub-trie manifest nodes)

Reply:
see pin_test.go tests for collections and directories..

Reply:
Added a test TestWalker in pin_test.go.. just for you :-)